Sap system restoration

CDimonaco opened the pull request:
# Description

Sap system  aggregate restoration

The flow is the same as registration, with the key difference that when the same conditions are met we fire a restoration event instead of registration one.

It could be possible that we have some leftovers in old database/application instances, in the case in which the user has not cleaned up the instances from an already deregistered sap system.

In that case the restoration process follows the same rules, without caring of leftovers, we assume that the user will take care of that, we guarantee that the new instances will follow the right path for restoration and correctness in our domain.


## How was this tested?

Automated tests

arbulu89 commented on lib/trento/domain/sap_system/sap_system.ex:
```diff
@@ -185,6 +242,27 @@ defmodule Trento.Domain.SapSystem do
+    |> Multi.execute(fn sap_system ->
+      maybe_emit_sap_system_restored_event(sap_system, instance)
+    end)
+    |> Multi.execute(&maybe_emit_sap_system_health_changed_event/1)
```
Why do you need this `multi`?
The restoration already changes the health, right?
And once it is restored, it wouldn't come from this flow
If it is for broadcasting messages, I think the restoration one should be enough

CDimonaco replied:
You're right, thanks for catching this

arbulu89 commented on test/trento/domain/sap_system/sap_system_test.exs:
```diff
@@ -1817,6 +1819,914 @@ defmodule Trento.SapSystemTest do
+          system_replication: "Secondary",
+          sid: db_sid
+        ),
+        build(
```
Do we need such a long test?
We don't need the application part to test this

arbulu89 commented on lib/trento/application/projectors/sap_system_projector.ex:
```diff
@@ -258,6 +284,21 @@ defmodule Trento.SapSystemProjector do
+        _,
+        _
+      ) do
+    sap_system = Repo.get!(SapSystemReadModel, sap_system_id)
```
The same

arbulu89 reviewed (commented):
I think that the restore events have data that wouldn't be changed.
Anyway, if you want to keep them, you don't need to get the data from the database in the projection `after_update` as you are getting updating all the data in the multi

I find the domain tests have too many preliminar conditions.
For example, for the database restoration, you don't need to setup a complete sap system.
The same for the sap system. You don't need to deregister a database instance to have a sap system deregistration. The sequence look extra complicated.

CDimonaco commented on lib/trento/domain/sap_system/sap_system.ex:
```diff
@@ -829,6 +939,27 @@ defmodule Trento.Domain.SapSystem do
     end
   end
 
+  defp maybe_emit_sap_system_restored_event(
```
Yes, I preferred to it on purpose for redability, just two duplicated lines

CDimonaco commented on lib/trento/domain/sap_system/events/sap_system_restored.ex:
```diff
@@ -0,0 +1,17 @@
+
+  defevent do
+    field :sap_system_id, Ecto.UUID
+    field :sid, :string
```
I was trying to mimic the registration as soon as possible, I noticed the updated event for ensa so I tought that was not really necessary, but we can add it if we want

CDimonaco commented on lib/trento/domain/sap_system/sap_system.ex:
```diff
@@ -185,6 +242,27 @@ defmodule Trento.Domain.SapSystem do
+    sap_system
+    |> Multi.new()
+    |> Multi.execute(fn sap_system ->
+      emit_application_instance_registered_or_application_instance_health_changed(
```
Yep, fixed

CDimonaco commented on lib/trento/domain/sap_system/sap_system.ex:
```diff
@@ -615,6 +693,26 @@ defmodule Trento.Domain.SapSystem do
+      ) do
+    %SapSystem{
+      sap_system
+      | sap_system_id: sap_system_id,
```
done thx

CDimonaco commented on lib/trento/domain/sap_system/events/database_restored.ex:
```diff
@@ -0,0 +1,15 @@
+
+  defevent do
+    field :sap_system_id, Ecto.UUID
+    field :sid, :string
```
You are right, I don't know what I was thinking 👀

CDimonaco commented on lib/trento/application/projectors/database_projector.ex:
```diff
@@ -279,6 +299,21 @@ defmodule Trento.DatabaseProjector do
+        _,
+        _
+      ) do
+    database = Repo.get!(DatabaseReadModel, sap_system_id)
```
Will remove sid

CDimonaco commented:
Thanks @arbulu89 for the fast feedback, addressed all the things and removed sid field from database/sap system restored events.

For the test "setup", I can short it if we want, I wanted to keep the scenario real even if the setup is much longer, but we could remove it if we want

arbulu89 reviewed (approved):
From Nuremberg! with love!!! :heart:
